### PR TITLE
HW5 - Layers - Ready to review

### PR DESCRIPTION
**Pull request by v4voloshyn**
app-architecture

### CODE REVIEW

**Comment by v4voloshyn** [comment]
It was interesting to learn how to rebuild your application using one of the architectural approaches.
I learned about _**Feature Sliced Design**_ and would probably use this approach in building larger applications than what I have now.
But for my application, I chose a **Modular approach** with minor customization.

The folder structure follows a **_Modular architecture approach_**, where the code is organized into modules: 
- Each module has its own folder containing components, its own API that its components can interact with, services, types, utilities, hooks, styles, and store slice specific to that module. 
- Pages folder created for collecting every single page with several modules that can interact with each other or complement each other. 
- There is a shared folder for components, API, services, types, utilities, global styles and global store as Redux, for example, that is used across multiple modules. Shared folder contains the Router also which interacts with the Database through API, fetching data using the `loader` method from `react-router-dom` and giving it its nested modules and components. 
- The `app` folder with `App.tsx` file serves as the entry point, and `main.ts` bootstraps the application. 

So, here Pros and Cons that i noticed and marked for myself:

### Pros:

- Reusability: Modules can be reused in different parts of the application or in other projects, saving development time and effort.
- Scalability: Easier scalability by adding or removing modules without impacting the entire system.
- Maintainability: Changes and bug fixes can be localized to specific modules, reducing the risk of unintended side effects and facilitating maintenance.
- Collaboration: Promotes team collaboration by allowing developers to work on different modules independently, enabling parallel development and faster iteration cycles.
- Testing: Modular systems are easier to test, as individual modules can be isolated and tested independently, simplifying unit testing and integration testing.

### Cons:

- Increased Complexity: Introducing modularity adds an extra layer of complexity to the system architecture, requiring careful management of module dependencies and communication.
- Overhead: The modular approach may introduce additional overhead in terms of communication and coordination between modules, necessitating proper design and architectural decisions.
- Performance Impact: Improperly designed modules can lead to performance overhead due to inter-module communication and data passing, necessitating optimization techniques.

Personally, for me, the biggest advantage of using this approach is that this structure promotes encapsulation, reusability, and clear separation of concerns.

P.S. This will be an interesting experience in terms of scaling the project by adding new modules, services, features, pages and so on. 😃

**Review by yulyaolshanska — APPROVED** [review]

File: src/modules/course-list/CourseList.module.tsx [hunk 1]
```diff
@@ -29,7 +29,7 @@ export const CourseList: FC = () => {
       <h1 className="course-list__title">Course List</h1>
       <div className="courses">
         {paginatedCourses.map((course) => {
-          return <CourseItem key={course.id} courseData={course} />;
+          return <CourseCard key={course.id} courseData={course} />;
```
Comment (Markusplay) [inline src/modules/course-list/CourseList.module.tsx:32]: To make code easier to read, you can remove `return` and add round brackets, like `course=>({CourseCard ...})`

File: src/modules/course-list/components/course-card/CourseCard.component.tsx [hunk 2]
```diff
@@ -2,18 +2,18 @@ import { FC, useEffect, useRef, useState } from 'react';
...
-import { CourseItemPreview } from '../../@types/types';
-import { Button } from '../common/UI/button/Button.component';
+import { CourseItemPreview } from '../../../../shared/@types/types';
+import { Button } from '../../../../shared/UI/button/Button.component';
+import { VideoPlayer } from '../../../../shared/components/video-player/VideoPlayer.component';
 import { SkillsList } from '../skills-list/SkillsList.component';
-import { VideoPlayer } from '../video-player/VideoPlayer.component';
 
-import './CourseItem.scss';
+import './CourseCard.scss';
```
Comment (Markusplay) [inline src/modules/course-list/components/course-card/CourseCard.component.tsx:10]: From the point of clean code principles, you need to specify for what component this interface is used. E.g: `ICourseCardProps`

File: src/modules/course-list/components/pagination/Pagination.component.tsx [hunk 3]
```diff
@@ -6,13 +6,13 @@ import { FIRST_PAGE, DEFAULT_ITEMS_PER_PAGE, PAGE_RANGE } from './constants';
 
 import './Pagination.scss';
 
-interface PaginationProps {
+interface Props {
```
Comment (Markusplay) [inline src/modules/course-list/components/pagination/Pagination.component.tsx:9]: Same problem as in CourseCard
https://stackoverflow.com/questions/65638991/interface-naming-convention-interface-name-clashes-with-component-name

File: src/modules/course-list/components/skills-list/SkillsList.component.tsx [hunk 4]
```diff
@@ -1,14 +1,14 @@
 import { FC } from 'react';
 
-import { ICourseMeta } from '../../@types/types';
+import { ICourseMeta } from '../../../../shared/@types/types';
 
 import './SkillsList.scss';
 
-interface SkillsListProps {
+interface Props {
```
Comment (Markusplay) [inline src/modules/course-list/components/skills-list/SkillsList.component.tsx:7]: I see that you have renamed your interfaces to `Props`. Can you explain it, or maybe provide link to conventions about interface naming

File: src/shared/UI/button/Button.component.tsx [hunk 5]
```diff
@@ -5,13 +5,13 @@ import { Spinner } from '../spinner/Spinner.component';
 
 import './Button.scss';
 
-interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
+interface Props extends ButtonHTMLAttributes<HTMLButtonElement> {
   buttonText?: string | ReactNode;
```
Comment (Markusplay) [inline src/shared/UI/button/Button.component.tsx:9]: buttonText can be ReactNode type ? Or rename buttonText varibale or delete ReactNode type

**Review by Markusplay — CHANGES_REQUESTED** [review]

File: src/modules/course-list/components/skills-list/SkillsList.component.tsx (hunk 4 above)
Comment (v4voloshyn) [reply src/modules/course-list/components/skills-list/SkillsList.component.tsx:7]: https://github.com/Genesis-Front-End-School/clean-code-principes-v4voloshyn/pull/4#discussion_r1184137641
Thats my convention of intarface naming :)

File: src/modules/course-list/CourseList.module.tsx (hunk 1 above)
Comment (v4voloshyn) [reply src/modules/course-list/CourseList.module.tsx:32]: It's ESLint plugin rule - to make arrow function return explicit.

File: src/shared/UI/button/Button.component.tsx (hunk 5 above)
Comment (v4voloshyn) [reply src/shared/UI/button/Button.component.tsx:9]: buttonText can be an icon in this case. But I think that you are right. I need to refactor it with startIcon prop and endIcon prop.

File: src/modules/course-list/components/pagination/Pagination.component.tsx (hunk 3 above)
Comment (v4voloshyn) [reply src/modules/course-list/components/pagination/Pagination.component.tsx:9]: All the renamed interfaces are internal interfaces. They are not imported or exported from anywhere. This is an internal contract of the component, so it is simply called ```Props```

**Comment by v4voloshyn** [comment]
@Markusplay I fixed some of your review points in my PR on HW6. Can you look at it and approve or make other notes on the current PR? [Link on commit](https://github.com/Genesis-Front-End-School/clean-code-principes-v4voloshyn/pull/8/commits/2228eaac35f414abbfbaf36e4e4380dc458ea240)

**Pull request merged**